ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+/* -*- Mode: C; indent-tabs-mode:nil -*- */
+/*
+ * Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+
+#include "mpi.h"
+#include "ompi/constants.h"
+#include "ompi/datatype/ompi_datatype.h"
+#include "ompi/mca/coll/base/coll_base_functions.h"
+#include "ompi/mca/coll/base/coll_base_util.h"
+#include "ompi/mca/coll/base/coll_tags.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/mca/pml/pml.h"
+#include "opal/util/bit_ops.h"
+#include "coll_acoll.h"
+#include "coll_acoll_utils.h"
+
+static inline int log_sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype,
+                                     int rank, int dim, int size, int sg_size, int cur_base,
+                                     int sg_start, struct ompi_communicator_t *comm,
+                                     mca_coll_base_module_t *module, ompi_request_t **preq,
+                                     int *nreqs)
+{
+    int msb_pos, sub_rank, peer, err;
+    int i, mask;
+    int end_sg, end_peer;
+
+    end_sg = sg_start + sg_size - 1;
+    if (end_sg >= size) {
+        end_sg = size - 1;
+    }
+    end_peer = (end_sg - cur_base) % sg_size;
+    sub_rank = (rank - cur_base + sg_size) % sg_size;
+
+    msb_pos = opal_hibit(sub_rank, dim);
+    --dim;
+
+    /* Receive data from parent in the sg tree. */
+    if (sub_rank > 0) {
+        assert(msb_pos >= 0);
+        peer = (sub_rank & ~(1 << msb_pos));
+        if (peer > end_peer) {
+            peer = (((peer + cur_base - sg_start) % sg_size) + sg_start);
+        } else {
+            peer = peer + cur_base;
+        }
+
+        err = MCA_PML_CALL(
+            recv(buff, count, datatype, peer, MCA_COLL_BASE_TAG_ALLGATHER, comm, MPI_STATUS_IGNORE));
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+
+    for (i = msb_pos + 1, mask = 1 << i; i <= dim; ++i, mask <<= 1) {
+        peer = sub_rank | mask;
+        if (peer >= sg_size) {
+            continue;
```

Comment:
Yes, we can change the `continue` to `break`.
Since `size `is `int`, `sub_rank`, `mask `and hence `peer `will be ensured to be within the positive `int `range.

No magic :slightly_smiling_face:, the logic is similar to the one in `mca_coll_basic_bcast_log_intra()`.